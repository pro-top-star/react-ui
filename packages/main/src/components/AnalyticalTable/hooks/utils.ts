export const makeTemplateColumns = (columns, resizedColumns) => {
  return columns
    .map((column) => {
      if (resizedColumns[column.id]) {
        return `${resizedColumns[column.id]}px`;
      }
      return `minmax(${column.minWidth}px, ${column.width || '1fr'})`;
    })
    .join(' ');
};
