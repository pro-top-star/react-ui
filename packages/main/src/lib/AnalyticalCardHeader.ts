import { AnalyticalCardHeader } from '../components/AnalyticalCard/header/AnalyticalCardHeader';

export { AnalyticalCardHeader };
