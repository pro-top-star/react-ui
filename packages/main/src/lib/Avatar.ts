import { Avatar } from '../webComponents/Avatar';

export { Avatar };
