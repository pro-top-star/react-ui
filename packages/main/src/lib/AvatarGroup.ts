import { AvatarGroup } from '../webComponents/AvatarGroup';
import type { AvatarGroupPropTypes } from '../webComponents/AvatarGroup';

export { AvatarGroup };
export type { AvatarGroupPropTypes };
