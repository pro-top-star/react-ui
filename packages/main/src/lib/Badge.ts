import { Badge } from '../webComponents/Badge';
import type { BadgePropTypes } from '../webComponents/Badge';

export { Badge };
export type { BadgePropTypes };
