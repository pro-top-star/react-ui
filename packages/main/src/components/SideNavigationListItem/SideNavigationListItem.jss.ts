import { CssSizeVariables } from '@ui5/webcomponents-react-base/lib/CssSizeVariables';
import { JSSTheme } from '../../interfaces/JSSTheme';

export const sideNavigationListItemStyles = ({ parameters }: JSSTheme) => ({
  listItem: {
    '&:active': {
      '--sapUiBaseText': parameters.sapUiListActiveTextColor,
      '& $icon, & $expandArrow': {
        '--sapUiContentNonInteractiveIconColor': parameters.sapUiListActiveTextColor
      }
    }
  },

  noIcons: {
    '& $text': {
      paddingLeft: '1rem'
    },
    '&[data-is-child] $text': {
      paddingLeft: '2rem'
    },
    boxSizing: 'border-box'
  },

  icon: {
    '--sapUiContentNonInteractiveIconColor': parameters.sapUiContentIconColor,
    width: CssSizeVariables.sapWcrSideNavigationItemIconSize,
    height: CssSizeVariables.sapWcrSideNavigationItemIconSize,
    padding: CssSizeVariables.sapWcrSideNavigationItemIconPadding,
    pointerEvents: 'none'
  },

  text: {},

  expandArrow: {
    '--sapUiContentNonInteractiveIconColor': parameters.sapUiContentIconColor,
    width: '0.875rem',
    height: '0.875rem',
    padding: CssSizeVariables.sapWcrSideNavigationItemExpandedArrowPadding,
    marginLeft: 'auto'
  },

  expanded: {
    '--ui5-listitem-border-bottom': 'none'
  },

  condensedExpandTriangle: {
    width: '0',
    height: '0',
    borderStyle: 'solid',
    borderWidth: '0 0 6px 6px',
    borderColor: `transparent transparent ${parameters.sapUiContentIconColor} transparent`,
    position: 'absolute',
    right: '0.125rem',
    bottom: '0.1875rem'
  }
});
