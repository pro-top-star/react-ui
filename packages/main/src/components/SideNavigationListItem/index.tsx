import '@ui5/webcomponents-icons/dist/icons/navigation-down-arrow.js';
import '@ui5/webcomponents-icons/dist/icons/navigation-right-arrow.js';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import { ContentDensity } from '@ui5/webcomponents-react/lib/ContentDensity';
import { CustomListItem } from '@ui5/webcomponents-react/lib/CustomListItem';
import { Icon } from '@ui5/webcomponents-react/lib/Icon';
import { List } from '@ui5/webcomponents-react/lib/List';
import { Popover } from '@ui5/webcomponents-react/lib/Popover';
import { PopoverVerticalAlign } from '@ui5/webcomponents-react/lib/PopoverVerticalAlign';
import { SideNavigationOpenState } from '@ui5/webcomponents-react/lib/SideNavigationOpenState';
import { StandardListItem } from '@ui5/webcomponents-react/lib/StandardListItem';
import { Text } from '@ui5/webcomponents-react/lib/Text';
import React, {
  Children,
  cloneElement,
  FC,
  forwardRef,
  ReactNode,
  ReactNodeArray,
  Ref,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from 'react';
import { createPortal } from 'react-dom';
import { createUseStyles, useTheme } from 'react-jss';
import { CommonProps } from '../../interfaces/CommonProps';
import { JSSTheme } from '../../interfaces/JSSTheme';
import { sideNavigationListItemStyles } from './SideNavigationListItem.jss';

export interface SideNavigationListItemProps extends CommonProps {
  icon?: string;
  text: string;
  id: number | string;
  children?: ReactNode | ReactNodeArray;
}

const useStyles = createUseStyles<JSSTheme, keyof ReturnType<typeof sideNavigationListItemStyles>>(
  sideNavigationListItemStyles,
  {
    name: 'SideNavigationListItem'
  }
);
const SideNavigationListItem: FC<SideNavigationListItemProps> = forwardRef(
  (props: SideNavigationListItemProps, ref: Ref<HTMLDivElement>) => {
    const { icon, text, id, children, tooltip, slot, className, style } = props;

    const [isExpanded, setExpanded] = useState(false);

    const handleToggleExpand = useCallback(() => {
      setExpanded(!isExpanded);
    }, [isExpanded, setExpanded]);

    const classes = useStyles();
    const theme = useTheme() as JSSTheme;

    const listItemClasses = StyleClassHelper.of(classes.listItem);
    if (theme.contentDensity === ContentDensity.Compact) {
      listItemClasses.put(classes.compact);
    }

    if (className) {
      listItemClasses.put(className);
    }

    if (isExpanded) {
      listItemClasses.put(classes.expanded);
    }

    const noIcons = props['noIcons'];
    if (noIcons) {
      listItemClasses.put(classes.noIcons);
    }

    const childCount = Children.count(children);
    const validChildren = Children.toArray(children).filter(Boolean);

    const isOpenStateExpanded = props['openState'] === SideNavigationOpenState.Expanded;

    useEffect(() => {
      if (validChildren.length) {
        const selectedElement = validChildren.find((child: any) => child.props.id === props['selectedId']);
        if (selectedElement) {
          setExpanded(isOpenStateExpanded);
        }
      }
    }, [props['selectedId'], id, children, setExpanded, isOpenStateExpanded]);

    const isSelfSelected = props['selectedId'] === id;
    const hasSelectedChild =
      !isOpenStateExpanded &&
      childCount > 0 &&
      !!validChildren.find((child: any) => child.props.id === props['selectedId']);

    const passThroughProps = usePassThroughHtmlProps(props);

    const customListItemCommonProps = {
      ref,
      className: listItemClasses.valueOf(),
      tooltip,
      slot,
      style,
      'data-id': id,
      'data-has-children': childCount > 0,
      'data-is-child': props['isChild'],
      ...passThroughProps
    };

    const popoverRef = useRef();

    const displayedIcon = useMemo(() => {
      return <Icon name={icon} className={classes.icon} />;
    }, [classes.icon, icon]);

    const handleOpenPopover = useCallback(
      (e) => {
        // @ts-ignore
        popoverRef.current.openBy(e.target);
      },
      [popoverRef.current]
    );

    if (isOpenStateExpanded) {
      return (
        <>
          <CustomListItem selected={isSelfSelected} {...customListItemCommonProps}>
            {icon && !noIcons && displayedIcon}
            {!icon && !noIcons && <span className={classes.icon} />}
            <Text className={classes.text}>{text}</Text>
            {childCount > 0 && (
              <span onClick={handleToggleExpand} className={classes.expandArrow}>
                <Icon name={isExpanded ? 'navigation-down-arrow' : 'navigation-right-arrow'} />
              </span>
            )}
          </CustomListItem>
          {isExpanded &&
            validChildren.map((child: any, index: number) => {
              const style = child.props.style || {};
              if (index !== childCount - 1) {
                style['--ui5-listitem-border-bottom'] = 'none';
              }

              return cloneElement(child, {
                icon: null,
                style,
                openState: props['openState'],
                selectedId: props['selectedId'],
                noIcons,
                isChild: true
              });
            })}
        </>
      );
    }

    return (
      <>
        <CustomListItem selected={isSelfSelected || hasSelectedChild} {...customListItemCommonProps}>
          {childCount > 0 ? (
            <span onClick={handleOpenPopover}>
              {displayedIcon}
              <div className={classes.condensedExpandTriangle} />
              {createPortal(
                <Popover ref={popoverRef} open={isExpanded} verticalAlign={PopoverVerticalAlign.Top}>
                  <List onItemClick={props['onListItemSelected']}>
                    <StandardListItem selected={isSelfSelected} data-id={id} tooltip={tooltip}>
                      {text}
                    </StandardListItem>
                    {validChildren.map((child: any, index) => (
                      <StandardListItem
                        selected={props['selectedId'] === child.props.id}
                        key={index}
                        data-id={child.props.id}
                        data-parent-id={id}
                        tooltip={child.props.tooltip || child.props.text}
                      >
                        {child.props.text}
                      </StandardListItem>
                    ))}
                  </List>
                </Popover>,
                document.body
              )}
            </span>
          ) : (
            displayedIcon
          )}
        </CustomListItem>
      </>
    );
  }
);

SideNavigationListItem.displayName = 'SideNavigationListItem';

SideNavigationListItem.defaultProps = {};

export { SideNavigationListItem };
