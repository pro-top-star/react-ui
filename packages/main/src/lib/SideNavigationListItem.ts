import { SideNavigationListItem } from '../components/SideNavigationListItem';

export { SideNavigationListItem };
