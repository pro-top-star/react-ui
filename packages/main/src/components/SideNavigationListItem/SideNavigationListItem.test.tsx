import { createPassThroughPropsTest, mountThemedComponent } from '@shared/tests/utils';
import { SideNavigationListItem } from '@ui5/webcomponents-react/lib/SideNavigationListItem';
import React from 'react';

describe('SideNavigationListItem', () => {
  test('Basic', () => {
    const wrapper = mountThemedComponent(<SideNavigationListItem text="Overview" icon="home" id="home" />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('custom class name and style', () => {
    const wrapper = mountThemedComponent(
      <SideNavigationListItem
        text="Overview"
        icon="home"
        id="home"
        className="myClass"
        style={{ backgroundColor: 'pink' }}
      />
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  createPassThroughPropsTest(SideNavigationListItem);
});
