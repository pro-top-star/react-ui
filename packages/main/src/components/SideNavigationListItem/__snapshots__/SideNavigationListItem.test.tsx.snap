// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNavigationListItem Basic 1`] = `
<ui5-li-custom
  class="SideNavigationListItem--listItem-"
  data-has-children="false"
  data-id="home"
  id="home"
  type="Active"
>
  <ui5-icon
    class="SideNavigationListItem--icon-"
    name="home"
  />
</ui5-li-custom>
`;

exports[`SideNavigationListItem custom class name and style 1`] = `
<ui5-li-custom
  class="SideNavigationListItem--listItem- myClass"
  data-has-children="false"
  data-id="home"
  id="home"
  style="background-color: pink;"
  type="Active"
>
  <ui5-icon
    class="SideNavigationListItem--icon-"
    name="home"
  />
</ui5-li-custom>
`;
